### PR TITLE
#3671 add defaultLocale and defaultCurrency to Channel model

### PR DESCRIPTION
**Pull request by axelvnk**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | #3671 
| License       | MIT

### CODE REVIEW

**Comment by axelvnk** [comment]
@pjedrzejewski I think we should add a listener on kernel.request that sets locale and currency for the session. But not sure where I should put this. CoreBundle would make the most sense I guess.

**Comment by axelvnk** [comment]
Testing in PHP7 fails, bin/behat --strict -f progress -p javascript_cached is not executed...

**Comment by pjedrzejewski** [comment]
I restarted the build, if this does not help, please rebase with master, there have been some tweaks to these scripts.

**Comment by axelvnk** [comment]
@pjedrzejewski Awesome, restarting the build seems to solve it.

File: src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Channel.orm.xml [hunk 1]
```diff
@@ -17,6 +17,12 @@
                                       http://doctrine-project.org/schemas/orm/doctrine-mapping.xsd">
 
     <mapped-superclass name="Sylius\Component\Core\Model\Channel" table="sylius_channel">
+        <many-to-one field="defaultLocale" target-entity="Sylius\Component\Locale\Model\LocaleInterface">
+            <join-column name="default_locale_id" referenced-column-name="id" nullable="true" on-delete="SET NULL"/>
+        </many-to-one>
+        <many-to-one field="defaultCurrency" target-entity="Sylius\Component\Currency\Model\CurrencyInterface">
+            <join-column name="default_currency_id" referenced-column-name="id" nullable="true" on-delete="SET NULL"/>
+        </many-to-one>
```
Comment (axelvnk) [inline src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Channel.orm.xml]: I made the fields nullable, although I'm not sure this is the way to go. How currencies/locales are set up right now doesn't allow me to make these not nullable. 

But perhaps we can fallback on the first locale/currency from the available ones that are set for the channel when the defaults are null. And if no locales/currencies are added we can fallback on the applications default locale/currency.

File: src/Sylius/Bundle/CoreBundle/Form/Type/ChannelType.php [hunk 2]
```diff
@@ -37,10 +37,18 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'label'    => 'sylius.form.channel.locales',
                 'multiple' => true
             ))
+            ->add('defaultLocale', 'sylius_locale_choice', array(
+                'label'    => 'sylius.form.channel.locale_default',
+                'multiple' => false
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Form/Type/ChannelType.php]: This is default value, no need to specify it here.

File: src/Sylius/Bundle/CoreBundle/Form/Type/ChannelType.php [hunk 3]
```diff
@@ -37,10 +37,18 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+            ))
             ->add('currencies', 'sylius_currency_choice', array(
                 'label'    => 'sylius.form.channel.currencies',
                 'multiple' => true
             ))
+            ->add('defaultCurrency', 'sylius_currency_choice', array(
+                'label'    => 'sylius.form.channel.currency_default',
+                'multiple' => false
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Form/Type/ChannelType.php]: Same as above. ^

File: src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Channel.orm.xml (hunk 1 above)
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Channel.orm.xml]: For now let's keep them nullable, but in the end, they should not be nullable. But that requires a bit more work, we will get there. :)

File: src/Sylius/Component/Core/Model/Channel.php [hunk 4]
```diff
@@ -69,6 +79,38 @@ public function __construct()
...
      */
+    public function getDefaultCurrency()
+    {
+        return $this->defaultCurrency;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setDefaultCurrency($defaultCurrency)
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Core/Model/Channel.php]: Missing typehints.

File: src/Sylius/Component/Core/Model/ChannelInterface.php [hunk 5]
```diff
@@ -38,4 +40,23 @@
...
+
+    /**
+     * @return LocaleInterface
+     */
+    public function getDefaultLocale();
+
+    /**
+     * @param CurrencyInterface $currency
+     */
+    public function setDefaultCurrency($currency);
```
Comment (pjedrzejewski) [inline src/Sylius/Component/Core/Model/ChannelInterface.php]: Missing typehints.

**Comment by pjedrzejewski** [comment]
Have a look at my comments, also, please squash commits, 11 is a bit too much for such simple addition. :) Thanks!

**Comment by axelvnk** [comment]
@pjedrzejewski done

**Comment by pjedrzejewski** [comment]
@axelvnk Sorry, I forgot to mention it earlier - please generate a migration for the database. Thanks for your time Axel!

**Comment by axelvnk** [comment]
@pjedrzejewski Hey is it possible someone forgot to generate migrations for their changes? Because I only have 5 query statements that differ from the currrent schema but the diff is generating 8 more than expected. See screenshot below :

screen shot 2015-12-08 at 10 10 50

**Comment by pjedrzejewski** [comment]
Can be but I see that you have associations which are not in core yet. :) Are you sure that these queries do not come from your custom code?

**Comment by axelvnk** [comment]
Omfg, you are right :') haha. Now that was dumb. I'll get rid of it and commit. Do you want me to squash again?

**Comment by pjedrzejewski** [comment]
@axelvnk Yes, please. :)

**Comment by axelvnk** [comment]
Okay should be all good now..

**Comment by pjedrzejewski** [comment]
Perfect, thanks again Axel! :+1:

File: src/Sylius/Bundle/CoreBundle/Behat/CoreContext.php [hunk 6]
```diff
@@ -72,7 +72,9 @@ public function storeHasDefaultConfiguration()
         $channel->setName('Default');
         $channel->setUrl('http://example.com');
         $channel->addCurrency($currency);
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Behat/CoreContext.php]: Btw. shouldn't currency and locale be added automatically when setting them as default?

**Pull request merged**